src/Microsoft.Data/BaseColumn.cs

```diff
@@ -0,0 +1,338 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+
+namespace Microsoft.Data
+{
+    /// <summary>
+    /// The base column type. All APIs should be defined here first
+    /// </summary>
+    public abstract partial class BaseColumn
+    {
+        public BaseColumn(string name, long length, Type type)
+        {
+            Length = length;
+            Name = name;
+            DataType = type;
+        }
+
+        private long _length;
+        public long Length
+        {
+            get => _length;
+            protected set
+            {
+                if (value < 0) throw new ArgumentOutOfRangeException();
+                _length = value;
+            }
+        }
+
+        public long NullCount { get; protected set; }
+
+        public string Name { get; set; }
+
+        public Type DataType { get; }
+
+        public virtual object this[long rowIndex]
+        {
+            get => GetValue(rowIndex);
+            set => SetValue(rowIndex, value);
+        }
+
+        protected virtual object GetValue(long rowIndex) => throw new NotImplementedException();
+        protected virtual void SetValue(long rowIndex, object value) => throw new NotImplementedException();
+
+        public virtual object this[long startIndex, int length]
+        {
+            get => throw new NotImplementedException();
+        }
+
+        /// <summary>
+        /// Clone column to produce a copy potentially changing the order by supplying mapIndices and an invert flag
+        /// </summary>
+        /// <param name="mapIndices"></param>
+        /// <param name="invertMapIndices"></param>
+        /// <returns></returns>
+        public virtual BaseColumn Clone(BaseColumn mapIndices = null, bool invertMapIndices = false) => throw new NotImplementedException();
+
+        public virtual BaseColumn Sort(bool ascending = true) => throw new NotImplementedException();
+
+        public virtual BaseColumn GetAscendingSortIndices() => throw new NotImplementedException();
+
+        protected delegate long GetBufferSortIndex(int bufferIndex, int sortIndex);
```

Comment:
You may want to make all these `internal`, as I don't think we'd want this implementation detail leaking in the public contract of this class.